src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -448,37 +454,27 @@ public virtual string CreateToken(string payload, SigningCredentials signingCred
 
-            return CreateTokenPrivate(JObject.Parse(payload), signingCredentials, encryptingCredentials, compressionAlgorithm, additionalHeaderClaims);
+            return CreateTokenPrivate(JObject.Parse(payload), signingCredentials, encryptingCredentials, compressionAlgorithm, additionalHeaderClaims, null);
         }
 
-        private string CreateTokenPrivate(JObject payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string compressionAlgorithm, IDictionary<string, object> additionalHeaderClaims)
+        private string CreateTokenPrivate(JObject payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string compressionAlgorithm, IDictionary<string, object> additionalHeaderClaims, string tokenType)
         {
-            if (additionalHeaderClaims != null && additionalHeaderClaims.Count > 0 && additionalHeaderClaims.Keys.Intersect(JwtTokenUtilities.DefaultHeaderParameters, StringComparer.OrdinalIgnoreCase).Any())
+            if (additionalHeaderClaims?.Count > 0 && additionalHeaderClaims.Keys.Intersect(JwtTokenUtilities.DefaultHeaderParameters, StringComparer.OrdinalIgnoreCase).Any())
                 throw LogHelper.LogExceptionMessage(new SecurityTokenException(LogHelper.FormatInvariant(LogMessages.IDX14116, nameof(additionalHeaderClaims), string.Join(", ", JwtTokenUtilities.DefaultHeaderParameters))));
 
-            var rawHeader = Base64UrlEncodedUnsignedJWSHeader;
-            // If there's no additional header claims to be added to the header and the token will be signed, try to retrieve a header value from the cache.
-            if ((additionalHeaderClaims == null || additionalHeaderClaims.Count == 0) && signingCredentials != null)
-            {
-                var header = CreateDefaultJWSHeader(signingCredentials);
-                rawHeader = Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(header.ToString(Formatting.None)));
-            } // Otherwise, if there is no outer JWT header, add additional header claims to this header.
-            else if (additionalHeaderClaims != null && additionalHeaderClaims.Count != 0)
-            {
-                JObject header;
-                if (signingCredentials != null)
-                    header = CreateDefaultJWSHeader(signingCredentials);
-                else
-                    header = new JObject
+            JObject header = null;
+            if (signingCredentials != null)
+                header = CreateDefaultJWSHeader(signingCredentials, tokenType);
+            else
+                header = new JObject
                     {
                         { JwtHeaderParameterNames.Alg, SecurityAlgorithms.None},
                     };
 
-                if (encryptingCredentials == null)
-                    header.Merge(JObject.FromObject(additionalHeaderClaims));
+            if (encryptingCredentials == null && additionalHeaderClaims != null && additionalHeaderClaims.Count != 0)
```

Comment:
little more clear / expressive to say this as 

`additionalHeaderClaims.Count > 0`

Reply:
I agree.